### PR TITLE
Add OAuth2 authentication (57)

### PR DESCRIPTION
**Pull request by jwhitlock**
## Proposed changes

To fix issue #57, add [OAuth2 authentication](https://tools.ietf.org/html/rfc6749) using [client credentials](https://oauth.net/2/grant-types/client-credentials/).

A new endpoint `/token` implements the OAuth2 client credentials grant process. A caller passes a ``client_id`` and a ``client_secret``, either in the form body or a header, and gets an OAuth2 token that expires. The token is implemented as a [JSON Web Token](https://en.wikipedia.org/wiki/JSON_Web_Token) or JWT, signed by the server (HMAC with SHA-256). When the token expires (default of 60 minutes), the client calls ``/token`` again.
    
All API endpoints now require a bearer token. The Swagger API docs already include a mechanism for authenticating in the browser.

There's a new database table ``api_client`` that stores the details for the API clients, along with an alembic migration. A script ``ctms/bin/client_credentials.py`` can create these. See below for an example.

There are new application settings:
    
 * ``CTMS_SECRET_KEY ``- A long string used in JWT signing. There is a default for development, and it should be unique for each deployment.
 * ``CTMS_TOKEN_EXPIRATION`` - How long, in seconds, that access tokens are valid. Default is 60 minutes.
 * ``CTMS_SERVER_PREFIX`` - The prefix of the server URL, defaults to the dev value of ``http://localhost:8000``. Currently used in the output of ``client_credentials.py``, but it may be handy for CSP or other uses.

## Types of changes

What types of changes does your code introduce?
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [guides](https://github.com/mozilla-it/ctms-api/tree/main/guides)
- [x] I have followed the [Mozilla Lean Data Policies](https://www.mozilla.org/en-US/about/policy/lean-data/)
- [x] Lint and tests pass both locally and on the cicd with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- *N/A* Any dependent changes have been merged and published in downstream modules

## Further comments

Sorry for the size of this PR. It is hard to do just a part of a security feature.

Here's how to use ``client_credentials.py`` in a local development environment:

```sh
make shell             # Enter a docker shell
poetry install         # Get ctms into Python path
alembic upgrade head   # Add the api_client table
ctms/bin/client_credentials.py test --email test@example.com
``` 

This will print something like:
```
Your OAuth2 client credentials are:

      client_id: id_test
  client_secret: secret_gPwcXQLSOXSYfpKxzjA5DBSQLDNL__MMkBBm2auWyxI

You can generate a token with an Authentication header:

  curl --user id_test:secret_gPwcXQLSOXSYfpKxzjA5DBSQLDNL__MMkBBm2auWyxI -F grant_type=client_credentials http://localhost:8000/token

or passing credentials in the form body:

  curl -v -F client_id=id_test -F client_secret=secret_gPwcXQLSOXSYfpKxzjA5DBSQLDNL__MMkBBm2auWyxI -F grant_type=client_credentials http://localhost:8000/token

The JSON response will have an access token, such as:

  {
    "access_token": "a_very_long_base64_string",
    "token_type": "bearer",
    "expires_in": 3600
  }

This can be used to access the API, such as:

  curl --oauth2-bearer a_very_long_base64_string http://localhost:8000/ctms?primary_email=test@example.com
```

You can then start the webserver (``make start`` in a different terminal) and enter the credentials into http://localhost:8000/docs.  The "Authorize" button in the Swagger docs is over by the right side:

CTMS authorize

You need to authorize before you can use the API endpoints, as hinted by the lock icons.

### CODE REVIEW

**Comment by jwhitlock** [comment]
Force-push to rebase on main and the package update PR

**Comment by bsieber-mozilla** [comment]
> Here's how to use client_credentials.py in a local development environment:
> 
> ```sh
> make shell             # Enter a docker shell
> poetry install         # Get ctms into Python path
> alembic upgrade head   # Add the api_client table
> ctms/bin/client_credentials.py test --email test@example.com```
 
When attempting to follow this:

- [x] make shell
- Results: SUCCESS

> => make shell
> Makefile:2: .env: No such file or directory
> Copying env.dist to .env...
> docker-compose -f ./docker-compose.yaml run  --rm web bash
> Starting ctms-api_postgres_1 ... done
> Creating ctms-api_web_run    ... done

- [x] poetry install
- Results: SUCCESS

> root@ba4aed097257:/app# poetry install
> Installing dependencies from lock file
> 
> 
> Package operations: 15 installs, 7 updates, 0 removals
> 
>   - Installing pycparser (2.20)
>   - Installing cffi (1.14.5)
>   - Installing pyasn1 (0.4.8)
>   - Installing argon2-cffi (20.1.0)
>   - Installing cryptography (3.4.6)
>   - Installing ecdsa (0.14.1)
>   - Updating gitpython (3.1.13 -> 3.1.14)
>   - Updating identify (1.5.13 -> 2.1.1)
>   - Installing mako (1.1.4)
>   - Updating pydantic (1.7.3 -> 1.8.1)
>   - Updating pygments (2.7.4 -> 2.8.1)
>   - Installing python-dateutil (2.8.1)
>   - Installing python-editor (1.0.4)
>   - Installing rsa (4.7.2)
>   - Installing alembic (1.5.6)
>   - Updating coverage (5.4 -> 5.5)
>   - Installing passlib (1.7.4)
>   - Updating pre-commit (2.10.1 -> 2.11.0)
>   - Installing python-jose (3.2.0)
>   - Installing python-multipart (0.0.5)
>   - Updating sphinx (3.4.3 -> 3.5.2)
>   - Installing sqlalchemy-utils (0.36.8)
>   - Installing ctms (0.4.0)
> 

- [ ] alembic upgrade head
- Results: FAIL

> root@ba4aed097257:/app# alembic upgrade head
> INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
> INFO  [alembic.runtime.migration] Will assume transactional DDL.
> INFO  [alembic.runtime.migration] Running upgrade  -> 9a91e36e6e6f, Init
> Traceback (most recent call last):
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
>     self.dialect.do_execute(
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute
>     cursor.execute(statement, parameters)
> psycopg2.errors.DuplicateTable: relation "emails" already exists
> 
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/opt/pysetup/.venv/bin/alembic", line 10, in <module>
>     sys.exit(main())
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/config.py", line 559, in main
>     CommandLine(prog=prog).main(argv=argv)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/config.py", line 553, in main
>     self.run_cmd(cfg, options)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/config.py", line 530, in run_cmd
>     fn(
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/command.py", line 294, in upgrade
>     script.run_env()
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/script/base.py", line 490, in run_env
>     util.load_python_file(self.dir, "env.py")
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/util/pyfiles.py", line 97, in load_python_file
>     module = load_module_py(module_id, path)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/util/compat.py", line 182, in load_module_py
>     spec.loader.exec_module(module)
>   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "migrations/env.py", line 82, in <module>
>     run_migrations_online()
>   File "migrations/env.py", line 76, in run_migrations_online
>     context.run_migrations()
>   File "<string>", line 8, in run_migrations
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/runtime/environment.py", line 813, in run_migrations
>     self.get_context().run_migrations(**kw)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/runtime/migration.py", line 560, in run_migrations
>     step.migration_fn(**kw)
>   File "/app/migrations/versions/20210224_9a91e36e6e6f_init.py", line 21, in upgrade
>     op.create_table(
>   File "<string>", line 8, in create_table
>   File "<string>", line 3, in create_table
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/operations/ops.py", line 1109, in create_table
>     return operations.invoke(op)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/operations/base.py", line 354, in invoke
>     return fn(self, operation)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/operations/toimpl.py", line 101, in create_table
>     operations.impl.create_table(table)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/ddl/impl.py", line 277, in create_table
>     self._exec(schema.CreateTable(table))
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/alembic/ddl/impl.py", line 146, in _exec
>     return conn.execute(construct, multiparams)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1011, in execute
>     return meth(self, multiparams, params)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/sql/ddl.py", line 72, in _execute_on_connection
>     return connection._execute_ddl(self, multiparams, params)
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1068, in _execute_ddl
>     ret = self._execute_context(
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context
>     self._handle_dbapi_exception(
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1510, in _handle_dbapi_exception
>     util.raise_(
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
>     raise exception
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
>     self.dialect.do_execute(
>   File "/opt/pysetup/.venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute
>     cursor.execute(statement, parameters)
> sqlalchemy.exc.ProgrammingError: (psycopg2.errors.DuplicateTable) relation "emails" already exists
> 
> [SQL: 
> CREATE TABLE emails (
> 	email_id UUID NOT NULL, 
> 	primary_email VARCHAR(255) NOT NULL, 
> 	basket_token VARCHAR(255), 
> 	sfdc_id VARCHAR(255), 
> 	mofo_id VARCHAR(255), 
> 	first_name VARCHAR(255), 
> 	last_name VARCHAR(255), 
> 	mailing_country VARCHAR(255), 
> 	email_format VARCHAR(1), 
> 	email_lang VARCHAR(5), 
> 	mofo_relevant BOOLEAN, 
> 	has_opted_out_of_email BOOLEAN, 
> 	unsubscribe_reason TEXT, 
> 	create_timestamp TIMESTAMP WITH TIME ZONE, 
> 	update_timestamp TIMESTAMP WITH TIME ZONE, 
> 	PRIMARY KEY (email_id), 
> 	UNIQUE (basket_token), 
> 	UNIQUE (primary_email)
> )
> 
> ]
> (Background on this error at: http://sqlalche.me/e/13/f405)


- [ ] ctms/bin/client_credentials.py test --email test@example.com
- Results: N/A

I got halted at the alembic upgrade head with the following err

**Comment by jwhitlock** [comment]
@bsieber-mozilla it sounds like you have a database that was populated before we added alembic. If you don't care about the contents, run ``make setup`` to re-initialize your database. If you do care, then ``alembic stamp 9a91e36e6e6f`` might do the right thing (create the revision table and say that you already applied the initial commit)

**Comment by bsieber-mozilla** [comment]
I just killed all running docker containers to try; still was having errors, re-cloned, killed all docker containers and I think I'm now in business, albeit noticed this on setup...

```
docker-compose exec postgres bash -c 'while !</dev/tcp/postgres/5432; do sleep 1; done'
bash: connect: Connection refused
bash: /dev/tcp/postgres/5432: Connection refused
docker-compose exec postgres dropdb postgres --user postgres

```

Will get back to this after standup

**Comment by bsieber-mozilla** [comment]
- [ ] alembic upgrade head
- Results: SUCCESS
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.

- [ ] root@774de875c06d:/app# ctms/bin/client_credentials.py test --email test@example.com
- Results: SUCCESS

----

- [ ] make start
- Results: Initially a failure, required a `make build`, then `make start`
- Also installed alembic and passlib to root, since both were complaining: "module not found" not sure if it was an order of operations or my operator err.

**Comment by bsieber-mozilla** [comment]
Screen Shot 2021-03-10 at 11 53 39 AM

**Comment by bsieber-mozilla** [comment]
Screen Shot 2021-03-10 at 11 57 21 AM
Screen Shot 2021-03-10 at 11 57 08 AM

**Comment by bsieber-mozilla** [comment]
Curl test:
```
| =>   curl --user id_test:secret_FzYSmBbV3g1slWUsMdVrd2F5Xs0hiQcwscahN6UHIck -F grant_type=client_credentials http://localhost:8000/token
{"access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJhcGlfY2xpZW50OmlkX3Rlc3QiLCJleHAiOjE2MTU0MDk5NDJ9.FXg2D_Gm8KREFusEByogxliNKIB9ASXivnRel91rgDs","token_type":"bearer","expires_in":3600}__
```

Ephemeral DB on local, so I imagine this is fine and not needed to be scrubbed.


Curl 2 test (form body):
```
| =>   curl -v -F client_id=id_test -F client_secret=secret_FzYSmBbV3g1slWUsMdVrd2F5Xs0hiQcwscahN6UHIck -F grant_type=client_credentials http://localhost:8000/token
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8000 (#0)
> POST /token HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 420
> Content-Type: multipart/form-data; boundary=------------------------eb58970bd4bd0aa7
> 
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< date: Wed, 10 Mar 2021 20:01:04 GMT
< server: uvicorn
< content-length: 200
< content-type: application/json
< 
* Connection #0 to host localhost left intact
{"access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJhcGlfY2xpZW50OmlkX3Rlc3QiLCJleHAiOjE2MTU0MTAwNjV9.DpnjQI78PUx4Ng3UnAGxX9zBgirUdFIbL8C1O4M2xaE","token_type":"bearer","expires_in":3600}* Closing connection 0
________________________________________________________________________________

```

File: guides/developer_setup.md [hunk 1]
```diff
@@ -194,6 +194,31 @@ python -m alembic upgrade head
 exit
 ```
 
+---
+## OAuth2 Client Credentials
```
Comment (bsieber-mozilla) [inline guides/developer_setup.md:198]: Thanks for this documentation!

**Review by bsieber-mozilla — APPROVED** [review]
I had a hiccup with my system, but don't believe it would be universal; it all appears to be running and functioning as expected.

LGTM, thanks for these efforts

**Comment by jwhitlock** [comment]
Rebased on current ``main``, no functional changes

**Pull request merged**